contrib/descend.man
.\" $Id$
.TH DESCEND 1 "31 March 1992"
.SH NAME 
descend \- walk directory tree and execute a command at each node
.SH SYNOPSIS 
.B descend
[
.B \-afqrv
]
.I command
[
.I directory
\&.\|.\|.
]
.SH DESCRIPTION
.B descend
walks down a directory tree and executes a command at each node.  It
is not as versatile as
.BR find (1),
but it has a simpler syntax.  If no
.I directory
is specified,
.B descend
starts at the current one.
.LP
Unlike
.BR find ,
.B descend
can be told to skip the special directories associated with RCS,
CVS, and SCCS.  This makes
.B descend
especially handy for use with these packages.  It can be used with
other commands too, of course.
.LP
.B descend
is a poor man's way to make any command recursive.  Note:
.B descend
does not follow symbolic links to directories unless they are
specified on the command line.
.SH OPTIONS
.TP 15
.B \-a
.I All.
Descend into directories that begin with '.'.
.TP
.B \-f
.I Force.
Ignore errors during descent.  Normally,
.B descend
quits when an error occurs.
.TP
.B \-q
.I Quiet.
Suppress the message `In directory
.IR directory '
that is normally printed during the descent.
.TP
.B \-r
.I Restricted.
Don't descend into the special directories 
.SB RCS,
.SB CVS,
.SB CVS.adm,
and
.SB SCCS.
.TP
.B \-v
.I Verbose.
Print 
.I command
before executing it.
.SH EXAMPLES
.TP 15
.B "descend ls"
Cheap substitute for `ls -R'.
.TP 15
.B "descend -f 'rm *' tree"
Strip `tree' of its leaves.  This command descends the `tree'
directory, removing all regular files.  Since 
.BR rm (1)
does not remove directories, this command leaves the directory
structure of `tree' intact, but denuded.  The
.B \-f
option is required to keep
.B descend
from quitting.  You could use `rm \-f' instead.
.TP 
.B "descend -r 'co RCS/*'" /project/src/
Check out every RCS file under the directory
.BR "/project/src" .
.TP
.B "descend -r 'cvs diff'"
Perform CVS `diff' operation on every directory below (and including)
the current one.
.SH DIAGNOSTICS
Returns 1 if errors occur (and the
.B \-f
option is not used).  Otherwise returns 0.
.SH SEE ALSO
.BR find (1),
.BR rcsintro (1),
.BR cvs (1),
.BR sccs (1)
.SH AUTHOR
Lowell Skoog
.br
Software Technology Group
.br
John Fluke Mfg. Co., Inc.
.SH BUGS
Shell metacharacters in
.I command
may have bizarre effects.  In particular, compound commands
(containing ';', '[', and ']' characters) will not work.  It is best
to enclose complicated commands in single quotes \(aa\ \(aa.


NEWS
Changes for NT (this title will be removed when this branch is merged
with the main trunk):

* The CVSROOT variable specification has changed to support various
"access methods."

* The "login" command no longer prompts the user for username and
hostname, since one will have to provide that information via the `-d'
flag or by setting CVSROOT.

Changes since 1.7:

* New "cvs annotate" command to display the last modification for each
line of a file, with the revision number, user checking in the
modification, and date of the modification.  For more information see
the `annotate' node in cvs.texinfo.

* Tag names containing only uppercase letters (no lowercase letters,
digits, or punctuation) are now reserved for CVS; it will not let you
add such tags (you can still use the tags if they were added with a
previous version of CVS).  This is so that CVS can in the future add
more special tag names analogous to BASE and HEAD.

* The cvsinit shell script has been replaced by a cvs init command.

* Added the patterns *.olb *.exe _$* *$ to default ignore list.

* There is now a $USER internal variable for *info files.

* There is no longer a separate `mkmodules' program; the functionality
is now built into `cvs'.  If upgrading an old repository, it is OK to
leave in the lines in the modules file which run mkmodules (the
mkmodules actions will get done twice, but that is harmless); you will
probably want to remove them once you are no longer using the old CVS.

* One can now specify user variables in *info files via the
${=varname} syntax; there is a -s global option to set them.  See the
Variables node in cvs.texinfo for details.

Changes from 1.6 to 1.7:

* The default ignore list has changed slightly: *.obj has been added
and CVS* has been changed to CVS CVS.adm.

* CVS now supports password authentication when accessing remote
repositories; this is useful for sites that can't use rsh (because of
a firewall, for example), and also don't have kerberos.  See node
"Password authenticated" (in "Remote repositories", in
doc/cvs.texinfo) for more details.  Note: This feature requires both
the client and server to be upgraded.

* Using the -kb option to specify binary files now works--most cases
did not work before.  See the "Binary files" section of
doc/cvs.texinfo for details.

* New developer communication features.  See the "Watches" section of
doc/cvs.texinfo for details.

* RCS keyword "Name" supported for "cvs update -r <tag>" and "cvs
checkout -r <tag>".

* If there is a group whose name matches a compiled in value which
defaults to "cvsadmin", only members of that group can use "cvs
admin".  This replaces the CVS_NOADMIN option.

* CVS now sets the modes of files in the repository based on the
CVSUMASK environment variable or a compiled in value defaulting to
002.  This way other developers will be able to access the files in
the repository regardless of the umask of the developer creating them.

* The command names in .cvsrc now match the official name of the
command, not the one (possibly an alias) by which it was invoked.  If
you had previously relied on "cvs di" and "cvs diff" using different
options, instead use a shell function or alias (for example "alias
cvsdi='cvs diff -u'").  You also can specify global CVS options (like
"-z") using the command name "cvs".

Changes from 1.5 to 1.6:

* Del updated the man page to include all of the new features
of CVS 1.6.

* "cvs tag" now supports a "-r | -D" option for tagging an already
tagged revision / specific revision of a file.

* There is a "taginfo" file in CVSROOT that supports filtering and
recording of tag operations.

* Long options support added, including --help and --version options.

* "cvs release" no longer cares whether or not the directory being
released has an entry in the `modules' file.

* The modules file now takes a -e option which is used instead of -o
for "cvs export".  If your modules file has a -o option which you want
to be used for "cvs export", change it to specify -e as well as -o.

* "cvs export" now takes a -k option to set RCS keyword expansion.
This way you can export binary files.  If you want the old behavior,
you need to specify -kv.

* "cvs update", "cvs rdiff", "cvs checkout", "cvs import", "cvs
release", "cvs rtag", and "cvs tag" used to take -q and -Q options
after the command name (e.g. "cvs update -q").  This was confusing
because other commands, such as "cvs ci", did not.  So the options
after the command name have been removed and you must now specify, for
example, "cvs -q update", which has been supported since CVS 1.3.

* New "wrappers" feature.  This allows you to set a hook which
transforms files on their way in and out of cvs (apparently on the
NeXT there is some particular usefulness in tarring things up in the
repository).  It also allows you to declare files as merge-by-copy
which means that instead of trying to merge the file, CVS will merely
copy the new version.  There is a CVSROOT/cvswrappers file and an
optionsl ~/.cvswrappers file to support this feature.

* You can set CVSROOT to user@host:dir, not just host:dir, if your
username on the server host is different than on the client host.

* VISUAL is accepted as well as EDITOR.

* $CVSROOT is expanded in *info files.

Changes from 1.4A2 to 1.5:

* Remote implementation.  This is very helpful when collaborating on a
project with someone across a wide-area network.  This release can
also be used locally, like other CVS versions, if you have no need for
remote access.

Here are some of the features of the remote implementation:
- It uses reliable transport protocols (TCP/IP) for remote repository
  access, not NFS.  NFS is unusable over long distances (and sometimes
  over short distances)
- It transfers only those files that have changed in the repository or
  the working directory.  To save transmission time, it will transfer
  patches when appropriate, and can compress data for transmission.

contrib/descend.sh
#! /bin/sh
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# descend - walk down a directory tree and execute a command at each node

fullname=$0
name=descend
usage="Usage: $name [-afqrv] command [directory ...]\n
\040\040-a\040\040All: descend into directories starting with '.'\n
\040\040-f\040\040Force: ignore errors during descent\n
\040\040-q\040\040Quiet: don't print directory names\n
\040\040-r\040\040Restricted: don't descend into RCS, CVS.adm, SCCS directories\n
\040\040-v\040\040Verbose: print command before executing it"

# Scan for options
while getopts afqrv option; do
    case $option in
	a)
	    alldirs=$option
	    options=$options" "-$option
	    ;;
	f)
	    force=$option
	    options=$options" "-$option
	    ;;
	q)
	    verbose=
	    quiet=$option
	    options=$options" "-$option
	    ;;
	r)
	    restricted=$option
	    options=$options" "-$option
	    ;;
	v)
	    verbose=$option
	    quiet=
	    options=$options" "-$option
	    ;;
	\?)
	    /usr/5bin/echo $usage 1>&2
	    exit 1
	    ;;
    esac
done
shift `expr $OPTIND - 1`

# Get command to execute
if [ $# -lt 1 ] ; then
    /usr/5bin/echo $usage 1>&2
    exit 1
else
    command=$1
    shift
fi

# If no directory specified, use '.'
if [ $# -lt 1 ] ; then
    default_dir=.
fi

# For each directory specified
for dir in $default_dir "$@" ; do

    # Spawn sub-shell so we return to starting directory afterward
    (cd $dir

	# Execute specified command
	if [ -z "$quiet" ] ; then
	    echo In directory `hostname`:`pwd`
	fi
	if [ -n "$verbose" ] ; then
	    echo $command
	fi
	eval "$command" || if [ -z "$force" ] ; then exit 1; fi

	# Collect dot file names if necessary
	if [ -n "$alldirs" ] ; then
	    dotfiles=.*
	else
	    dotfiles=
	fi

	# For each file in current directory
	for file in $dotfiles * ; do

	    # Skip '.' and '..'
	    if [ "$file" = "." -o "$file" = ".." ] ; then
		continue
	    fi

	    # If a directory but not a symbolic link
	    if [ -d "$file" -a ! -h "$file" ] ; then

		# If not skipping this type of directory
		if [ \( "$file" != "RCS" -a \
			"$file" != "SCCS" -a \
			"$file" != "CVS" -a \
			"$file" != "CVS.adm" \) \
			-o -z "$restricted" ] ; then

		    # Recursively descend into it
		    $fullname $options "$command" "$file" \
		    || if [ -z "$force" ] ; then exit 1; fi
		fi

	    # Else if a directory AND a symbolic link
	    elif [ -d "$file" -a -h "$file" ] ; then

		if [ -z "$quiet" ] ; then
		    echo In directory `hostname`:`pwd`/$file: symbolic link: skipping
		fi
	    fi
	done
    ) || if [ -z "$force" ] ; then exit 1; fi
done
